### PR TITLE
created a dotnet core csharp template

### LINKED ISSUE #2858: add C# service template
# This is a Feature Proposal

## Description

More about C# programming model http://docs.aws.amazon.com/lambda/latest/dg/dotnet-programming-model.html

### PR DESCRIPTION
**Pull request by jimmyp**
## What did you implement:

Added csharp dotnetcore template support

Closes #2858

## How did you implement it:

Added a very simple csharp template with an integration test.

The template contains only the http handler the integration test requires. I did add a package for json serialisation and deserialisation otherwise the input and output of the handler would be just streams.

## How can we verify it:

I verified it by creating and running the unit test. I wasn't sure how to create a local version that included my template and do any manual testing.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Enable ["Allow edits from maintainers"](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) for this PR
- [x] Change ready for review message below

***Is this ready for review?:*** YES

### CODE REVIEW

File: lib/plugins/create/create.js [hunk 1]
```diff
@@ -11,6 +11,7 @@ const validTemplates = [
   'aws-java-maven',
   'aws-java-gradle',
   'aws-scala-sbt',
+  'aws-dotnet-csharp',
```
Comment (mthenw) [inline lib/plugins/create/create.js]: `aws-csharp` is fine

Comment (jimmyp) [inline lib/plugins/create/create.js]: 👍

**Comment by jimmyp** [comment]
will do @pmuens 

btw how do I check the coverage? And what output am I expecting from linting... all I get locally is this:

```~/dev/serverless [csharp_dotnetcore_template|✔]
23:27 $ npm run lint

> serverless@1.3.0 lint /Users/jim/dev/serverless
> eslint .
```

is that saying no lint errors?

File: lib/plugins/create/templates/aws-csharp/build.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+#build handlers
+dotnet restore
+dotnet publish -c release
+
+#install zip
+apt-get -qq update
+apt-get -qq -y install zip
```
Comment (mthenw) [inline lib/plugins/create/templates/aws-csharp/build.sh]: I think we can skip those two lines and just assume that zip is already installed.

File: lib/plugins/create/templates/aws-csharp/serverless.yml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+#          - - "arn:aws:s3:::"
+#            - "Ref" : "ServerlessDeploymentBucket"
+
+# you can define service wide environment variables here
+#  environment:
+#    variable1: value1
+
+# you can add packaging information here
+package:
+ artifact: bin/Release/netcoreapp1.0/publish/deploy-package.zip
```
Comment (adamhathcock) [inline lib/plugins/create/templates/aws-csharp/serverless.yml]: can do the `-o` publish on `dotnet publish` to specify an output directory

Comment (jimmyp) [inline lib/plugins/create/templates/aws-csharp/serverless.yml]: I can definitely customise it. The path is pretty standard for the output of a dotnet build.

I can change the output path of the build or I could change the output of zipping?

What were you thinking?

File: lib/plugins/create/templates/aws-csharp/build.sh (hunk 2 above)
Comment (jimmyp) [inline lib/plugins/create/templates/aws-csharp/build.sh]: Wouldn't that be nice, but unfortunately zip isn't installed on the microsoft supplied dotnet docker images like the one I've used `microsoft/dotnet:1.0.1-sdk-projectjson`

```
...
publish: Published to /app/bin/release/netcoreapp1.0/publish
Published 1/1 projects successfully
/app/bin/Release/netcoreapp1.0/publish /app
./build.sh: line 13: zip: command not found
```

File: docs/providers/aws/examples/hello-world/aws-csharp/README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
+<!--
+title: Hello World Node.js Example
+menuText: Hello World Node.js Example
+description: Create a Node.js Hello World Lambda function
+layout: Doc
+-->
+
+<!-- DOCS-SITE-LINK:START automatically generated  -->
+### [Read this on the main serverless docs site](https://www.serverless.com/framework/docs/providers/aws/examples/hello-world/node/)
```
Comment (pmuens) [inline docs/providers/aws/examples/hello-world/aws-csharp/README.md]: This link needs to be updated

File: docs/providers/aws/examples/hello-world/aws-csharp/README.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
+<!--
+title: Hello World Node.js Example
```
Comment (pmuens) [inline docs/providers/aws/examples/hello-world/aws-csharp/README.md]: These meta information needs to be updated

File: docs/providers/aws/examples/hello-world/aws-csharp/README.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+<!-- DOCS-SITE-LINK:START automatically generated  -->
+### [Read this on the main serverless docs site](https://www.serverless.com/framework/docs/providers/aws/examples/hello-world/node/)
+<!-- DOCS-SITE-LINK:END -->
+
+# Hello World CSharp Example
+
+Make sure `serverless` is installed. [See installation guide](../../../guide/installation.md).
+
+## 1. Create a service
+`serverless create --template aws-csharp --path myService` or `sls create --template aws-ncsharp --path myService`, where 'myService' is a new folder to be created with template service files.  Change directories into this new folder.
```
Comment (pmuens) [inline docs/providers/aws/examples/hello-world/aws-csharp/README.md]: `aws-ncsharp` --> `aws-csharp`

File: docs/providers/aws/examples/hello-world/csharp/README.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
+<!--
+title: Hello World Node.js Example
```
Comment (pmuens) [inline docs/providers/aws/examples/hello-world/csharp/README.md]: Same updates need to be applied here

File: docs/providers/aws/examples/hello-world/csharp/README.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+menuText: Hello World Node.js Example
+description: Create a Node.js Hello World Lambda function
+layout: Doc
+-->
+
+<!-- DOCS-SITE-LINK:START automatically generated  -->
+### [Read this on the main serverless docs site](https://www.serverless.com/framework/docs/providers/aws/examples/hello-world/node/)
```
Comment (pmuens) [inline docs/providers/aws/examples/hello-world/csharp/README.md]: This needs to be updated as well

File: docs/providers/aws/examples/hello-world/csharp/README.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+<!-- DOCS-SITE-LINK:END -->
+
+# Hello World CSharp Example
+
+Make sure `serverless` is installed. [See installation guide](../../../guide/installation.md).
+
+## 1. Create a service
+`serverless create --template aws-csharp --path myService` or `sls create --template aws-ncsharp --path myService`, where 'myService' is a new folder to be created with template service files.  Change directories into this new folder.
```
Comment (pmuens) [inline docs/providers/aws/examples/hello-world/csharp/README.md]: `aws-ncsharp` --> `aws-csharp`

File: docs/providers/aws/examples/hello-world/aws-csharp/README.md (hunk 5 above)
Comment (jimmyp) [inline docs/providers/aws/examples/hello-world/aws-csharp/README.md]: thanks

File: docs/providers/aws/examples/hello-world/aws-csharp/README.md (hunk 4 above)
Comment (jimmyp) [inline docs/providers/aws/examples/hello-world/aws-csharp/README.md]: thanks

**Comment by jimmyp** [comment]
Just mangled the PR with a rebase, please stand by :(

**Comment by jimmyp** [comment]
Ok, thanks for your patience. Commit history is all fixed and matches the contribution guidelines.

I've fixed the copy paste mistakes in the doco. Two items left outstanding. First is the conversation around installing zip as part of the build script, which is needed for the integration test. Second is the output path of the dotnet build and the zip package artefact. Happy to take advice on either.

**Comment by pmuens** [comment]
Thanks for the updates @jimmyp! 👍 

Could you also add the template to the services.md file (in /docs/providers/aws/guide).

Thanks in advance!

**Comment by jimmyp** [comment]
@pmuens done. I had a scan through the rest of the doco but didn't find anything else obvious to update?

**Comment by pmuens** [comment]
@jimmyp thanks for the update!

Have you pushed your changes to this branch? Looked through it but couldn't find an update for the `services.md` file.

**Comment by jimmyp** [comment]
@pmuens ummm... so... I *have* pushed my changes, you can see them in the branch on my fork https://github.com/jimmyp/serverless/commits/csharp_dotnetcore_template but that last commit is not park of this PR?

I've never come across this before using github. Any ideas?

**Comment by jimmyp** [comment]
@pmuens I switched the PR base to another random branch and back to master and it seems to have picked up the last commit now. Should be good to go.

**Comment by jimmyp** [comment]
🎉

**Pull request merged**